src/components/App/result-cache.ts

```diff
@@ -0,0 +1,68 @@
+import { EncoderState } from '../../codecs/encoders';
+import { shallowEqual } from '../../lib/util';
+import { SourceImage } from '.';
+import { PreprocessorState } from '../../codecs/preprocessors';
+
+import * as identity from '../../codecs/identity/encoder';
+
+interface CacheResult {
+  preprocessed: ImageData;
+  bmp: ImageBitmap;
+  file: File;
+}
+
+interface CacheEntry extends CacheResult {
+  preprocessorState: PreprocessorState;
+  encoderState: EncoderState;
+  source: SourceImage;
+}
+
+const SIZE = 5;
```

Comment:
Given how big images can be, I went for a small number here. Feels like we needed at least 4 to allow toggling an option on each side.